src/main/java/org/ice4j/util/AsyncQueueHandler.java

```diff
@@ -0,0 +1,459 @@
+    }
+
+    /**
+     * Attempts to cancel currently running reader. Assuming called when
+     * lock on {@link #syncRoot} is already taken
+     * @param mayInterruptIfRunning indicates if {@link #reader} allowed
+     * to be interrupted if running
+     */
+    private void cancel(boolean mayInterruptIfRunning)
+    {
+        if (delayedFuture != null)
+        {
+            delayedFuture.cancel(mayInterruptIfRunning);
+            delayedFuture = null;
+        }
+
+        if (readerFuture != null)
+        {
+            readerFuture.cancel(mayInterruptIfRunning);
+            readerFuture = null;
+        }
+    }
+
+    /**
+     * A simple interface to handle enqueued {@link T} items.
+     * @param <T> the type of the item.
+     */
+    public interface Handler<T>
+    {
+        /**
+         * Does something with an item.
+         * @param item the item to do something with.
+         */
+        void handleItem(T item);
+
+        /**
+         * Specifies max number {@link #handleItem(T)} invocation
+         * per {@link #perNanos()}
+         * @return positive number of allowed handled items in case of
+         * throttling must be enabled.
+         */
+        default long maxHandledItems() {
+            return -1;
+        }
+
+        /**
+         * Specifies time interval in nanoseconds for {@link #maxHandledItems()}
+         * @return positive nanoseconds count in case of throttling must be
+         * enabled.
+         */
+        default long perNanos() {
+            return -1;
+        }
+
+        /**
+         * Specifies the number of items allowed to be handled sequentially
+         * without yielding control to executor's thread. Specifying positive
+         * number will allow other possible queues sharing same
+         * {@link ExecutorService} to process their items.
+         * @return positive value to specify max number of sequentially handled
+         * items which allows implementation of cooperative multi-tasking
+         * between different {@link AsyncQueueHandler<T>} sharing
+         * same {@link ExecutorService}.
+         */
+        default long maxSequentiallyHandledItems()
+        {
+            return -1;
+        }
+    }
+
+    /**
+     * Helper class to calculate throttle delay when throttling is enabled.
+     */
+    private static final class ThrottleCalculator<T>
+    {
+        /**
+         * The number of {@link T}s already processed during the current
+         * <tt>perNanos</tt> interval.
+         */
+        private long itemsHandledWithinInterval = 0;
+
+        /**
+         * The time stamp in nanoseconds of the start of the current
+         * <tt>perNanos</tt> interval.
+         */
+        private long intervalStartTimeNanos = 0;
+
+        /**
+         * {@link Handler <T>} instance which provide throttling
+         * configuration
+         */
+        private final Handler<T> handler;
+
+        ThrottleCalculator(Handler<T> handler)
```

Comment:
i'm still getting stuck on the need for the throttling.  i disagree about the amount of complexity it adds, because although the mechanism itself may be somewhat simple there's an extra future that needs to be tracked and synchronized correctly, as well as logic that needs to be understood and properly dealt with if any changes are to be made (i took a look at removing the logic for this, the file has over 25% fewer lines of code with it gone). since it is not being used i'm feeling like we should just remove this.  to me, it even feels unlikely to use the maxSequential field.

i know you are just porting over what's there, just my thoughts.

Reply:
ok, removing. Just curios what is the timeline of new bridge arriving, so I will not need this functionality?

Reply:
> maybe i misunderstood something, i didn't think this was actually being used anywhere (implemented, yes, but was the functionality actually in use?)

I personally didn't check if it's invoked in runtime at the moment of implementing it in `PacketQueue` and for my use case of audio only conferences it doesn't not happen, because the only place this invoked is from [VideoMediaStreamImpl](https://github.com/jitsi/libjitsi/blob/6c0a31d6193dcd2b012ae3ced656ad403403d9ff/src/org/jitsi/impl/neomedia/VideoMediaStreamImpl.java#L694).

But since you've asked, I've just checked conference with video stream under debugger and it looks like by default this code is not invoked, because `maxBandwith` is `2^31-1`.